# Big Data Project
## Expected outcome

In this notebook we will simulate expected outcome indicators

In [2]:
import os
import os.path
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from colorama import Fore

pathori = "../Data/Raw/Main/"
pathtemp = "../Data/Interim/"
pathdest = "../Data/Processed/"
filetemp = "main_competitions.csv"
filedest = "main_competitions.csv"

In [4]:
# Open dataset
url = pathtemp + filetemp
parse_dates=["Date"]
ds = pd.read_csv(url, parse_dates=parse_dates, index_col=False)

In [5]:
display(ds.dropna().sample(10))

Country Competition     Season  Div       Date   HomeTeam  \
95881     Spain     Segunda  2005-2006  SP2 2006-04-29    Levante   
26609   England     League2  2006-2007   E3 2007-10-02    Grimsby   
92097     Spain     Primera  2009-2010  SP1 2009-11-21  La Coruna   
96958     Spain     Segunda  2009-2010  SP2 2009-10-10      Celta   
97485     Spain     Segunda  2010-2011  SP2 2010-11-15   Alcorcon   
50092   Germany       Liga1  2009-2010   D1 2009-08-15  Stuttgart   
90781     Spain     Primera  2005-2006  SP1 2006-02-04      Betis   
88105  Scotland     Premier  2009-2010  SC0 2010-01-23    Rangers   
19158   England     League1  2006-2007   E2 2006-12-23   Tranmere   
49087   Germany       Liga1  2005-2006   D1 2006-03-25   Duisburg   

            AwayTeam FTR   WHH   WHD   WHA   SBH   SBD   SBA   IWH  IWD   IWA  \
95881       Hercules   H  1.70  3.00  5.00  1.67  3.00  5.50  1.65  3.1  4.80   
26609    Bristol Rvs   H  2.37  3.10  2.62  2.40  3.20  2.70  2.40  3.0  2.60   
92097     Ath Madrid   H  2.40  3.20  2.75  2.35  3.30  2.70  2.40  3.2  2.85   
96958       Albacete   D  2.05  3.10  3.50  1.95  3.10  3.75  2.00  3.0  3.30   
97485        Cordoba   H  1.91  3.10  4.00  2.00  3.10  3.60  2.10  3.0  3.10   
50092       Freiburg   H  1.40  4.00  6.50  1.40  4.00  7.50  1.45  4.0  7.00   
90781        Sevilla   H  2.40  2.80  2.87  2.50  3.10  2.80  2.55  2.8  2.85   
88105         Hearts   D  1.30  4.33  8.00  1.30  4.50  9.00  1.37  4.2  7.50   
19158     Gillingham   A  1.72  3.25  4.20  1.73  3.40  4.33  1.75  3.2  4.00   
49087  Bayern Munich   A  6.50  3.60  1.44  6.50  4.33  1.44  5.50  3.8  1.50   

        GBH   GBD   GBA  
95881  1.65  3.10  5.00  
26609  2.40  3.20  2.65  
92097  2.45  3.25  2.75  
96958  1.95  3.10  3.75  
97485  2.00  3.00  3.70  
50092  1.45  4.00  7.25  
90781  2.35  3.15  2.90  
88105  1.33  4.50  8.00  
19158  1.70  3.40  4.50  
49087  8.00  4.00  1.40

**Success of a betting house**

In fact this indicator is the **'Accuracy'** because it is 'Success (TP and TN) related to total matches.

We expect to define a function like this: **accuracy (match, bethouse, scope)**

With params:
- match: Full information (register) of a match
- bethouse: Abrevation of the Bet House. Values: WH, SB, IW, GB
- scope: Columns to filter information. Values:  Country, Competition, Season, Team or combination of them.

This function will calculate accuracy = success/matches of a concret Bet House in the selected scope and in reference to the desired match.

Example: 
match = 

 	Country 	Competition 	Season 	Div 	Date 	HomeTeam 	AwayTeam 	FTR 	WHH 	WHD 	WHA 	SBH 	SBD 	SBA 	IWH 	IWD 	IWA 	GBH 	GBD 	GBA
    
0 	Belgium 	JupilerLeague 	2003-2004 	B1 	2003-08-08 	Club Brugge 	Genk 	H 	NaN 	NaN 	NaN 	1.44 	3.75 	6.5 	1.45 	3.8 	5.4 	1.4 	3.8 	6.85

result = accuracy(match, 'WH', ['Spain'])

result = 0.6


**Divergence**

In a individual match divergence it's a maesure of the difference between Bet House Quotes.

We expect to define a function like this: divergence(match)

We will use this function following these steps:

1. Filter matches of this week. Optional we can filter also by country or competition

2. Map rows calculating divergence

3. Reduce result amb obtain top N divergence values 

Result will be a Dataframe with the N rows with highest divergence, that they are the ones we need to pay attention to.


** Prediction Model **

Probably we will use a Classification Model, not selected yet, but that obtains a percentage of probability of win, draw or lose.

Example:
features = ['Country','Competition','Season','HomeTeam','AwayTeam','WHH','WHD','WHA','SBH','SBD','SBA','IWH','IWD','IWA','GBH','GBD','GBA']

label = 'FTR'

model = Model.trail(features, label)

prediction = model.predict(features)

Result:

Win: 'Belgium JupilerLeague 2003-2004 B1 2003-08-08 Club Brugge Genk  ...', 0.60

Drop: 'Belgium JupilerLeague 2003-2004 B1 2003-08-08 Club Brugge Genk  ...', 0.30
